nupic/research/frameworks/continuous_learning/k_winners.py

```diff
@@ -0,0 +1,367 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2020, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+import abc
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+import nupic.torch.functions as F
+from nupic.torch.duty_cycle_metrics import binary_entropy, max_entropy
+
+
+def update_boost_strength(m):
+    """Function used to update KWinner modules boost strength. This is typically done
+    during training at the beginning of each epoch.
+
+    Call using :meth:`torch.nn.Module.apply` after each epoch if required
+    For example: ``m.apply(update_boost_strength)``
+
+    :param m: KWinner module
+    """
+    if isinstance(m, KWinnersBase):
+        m.update_boost_strength()
+
+
+def new_epoch(m):
+    if isinstance(m, KWinnersBase):
+        m.get_new_epoch()
+
+
+def per_epoch(m, bpe):
+    if isinstance(m, KWinnersBase):
+        m.per_epoch = bpe
+
+
+class KWinnersBase(nn.Module, metaclass=abc.ABCMeta):
```

Comment:
Not clear why all this is here. If something different is happening, it should be named differently and the differences should be spelled out in the docstring.